lib/std/crypto/phc_encoding.zig

```diff
@@ -0,0 +1,391 @@
+const version_param_name = "v";
+const params_delimiter = ",";
+const kv_delimiter = "=";
+
+pub const Error = std.crypto.errors.EncodingError || error{NoSpaceLeft};
+
+const B64Decoder = std.base64.standard_no_pad.Decoder;
+const B64Encoder = std.base64.standard_no_pad.Encoder;
+
+/// A wrapped binary value whose maximum size is `max_len`.
+///
+/// This type must be used whenever a binary value is encoded in a PHC-formatted string.
+/// This includes `salt`, `hash`, and any other binary parameters such as keys.
+///
+/// Once initialized, the actual value can be read with the `unwrap()` function.
+pub fn BinValue(comptime max_len: usize) type {
+    return struct {
+        const Self = @This();
+        const capacity = max_len;
+        const max_encoded_length = B64Encoder.calcSize(max_len);
+
+        buf: [max_len]u8 = undefined,
+        len: usize = 0,
+
+        /// Wrap an existing byte slice
+        pub fn fromSlice(slice: []const u8) Error!Self {
+            if (slice.len > capacity) return Error.NoSpaceLeft;
+            var bin_value: Self = undefined;
+            mem.copy(u8, &bin_value.buf, slice);
+            bin_value.len = slice.len;
+            return bin_value;
+        }
+
+        /// Return the slice containing the actual value.
+        pub fn unwrap(self: Self) []const u8 {
+            return self.buf[0..self.len];
+        }
+
+        fn fromB64(self: *Self, str: []const u8) !void {
+            const len = B64Decoder.calcSizeForSlice(str) catch return Error.InvalidEncoding;
+            if (len > self.buf.len) return Error.NoSpaceLeft;
+            B64Decoder.decode(&self.buf, str) catch return Error.InvalidEncoding;
+            self.len = len;
+        }
+
+        fn toB64(self: Self, buf: []u8) ![]const u8 {
+            const value = self.unwrap();
+            const len = B64Encoder.calcSize(value.len);
+            if (len > buf.len) return Error.NoSpaceLeft;
+            return B64Encoder.encode(buf, value);
+        }
+    };
+}
+
+/// Deserialize a PHC-formatted string into a structure `HashResult`.
+///
+/// Required field in the `HashResult` structure:
+///   - `alg_id`: algorithm identifier
+/// Optional, special fields:
+///   - `alg_version`: algorithm version (unsigned integer)
+///   - `salt`: salt
+///   - `hash`: output of the hash function
+///
+/// Other fields will also be deserialized from the function parameters section.
+pub fn deserialize(comptime HashResult: type, str: []const u8) Error!HashResult {
+    var out: HashResult = undefined;
+    var it = mem.split(u8, str, fields_delimiter);
+    var set_fields: usize = 0;
+
+    // FIXME
+    // macOS x86_64
+    // ./build/release/bin/zig test ./lib/std/crypto/phc_encoding.zig -OReleaseFast --test-filter "encoding"
+    // .{
+    //     .str = "$scrypt$ln=15,r=8,p=1",
+    //     .HashResult = struct { alg_id: []const u8, alg_version: ?u30, ln: u6, r: u30, p: u30 },
+    // },
+    // out.alg_version != null (`00`)
```

Comment:
Is there an issue here?

Reply:
```zig
var out: HashResult = undefined;
```

On macOS arm `out.alg_version` will be `null`, but on macOS x86_64 it will be `00`. It was unexpected for me. I can't say for sure if it is a bug or a future.

```zig
if (@hasField(HashResult, "alg_version")) {
    if (@typeInfo(@TypeOf(out.alg_version)) == .Optional) {
        out.alg_version = null;
    }
}
```

It is a fix for `00`. But maybe it have to be `null` on all platforms and it is a bug inside zig?

Reply:
Removed comments, mark resolved

Reply:
`var out: HashResult = undefined;`

`out` is `undefined`, so optional pointers are not assumed to be set to `null`.

Optionals in `HashResult` can have a `null` default value, but as demonstrated here, it is easy to forget. As deserialization is expected to not leave anything uninitialized, instead of assigning `undefined` to `HashResult` and making `alg_version` a special case, maybe use `mom::zeroes()` instead. This will prevent this kind of issue from happening again if more optional values are required.